[AMD-ZENDNN] Modifications to support eager op rewrite for AMD CPU Inference

sanadani opened the pull request:
This pull request introduces support to re-write eager ops with ZenOps.
All these code changes are under AMD_ZENDNN flag which is enabled for linux builds only.

Authors:
Aakar Dwivedi ( aakar.dwivedi@amd.com )
Avinash-Chandra Pandey ( avinash-chandra.pandey@amd.com )
Chandra Kumar Ramasamy ( chandrakumar.ramasamy@amd.com )
Savan Anadani ( savan.anadani@amd.com )

Signed-off-by: sanadani <savan.anadani@amd.com>

sanadani committed:
[AMD-ZENDNN] Modifications to support eager op rewrite for AMD CPU Inference
This pull request introduces support to re-write eager ops with ZenOps.
All these code changes are under AMD_ZENDNN flag which is enabled for linux builds only.

Authors:
  Aakar Dwivedi ( aakar.dwivedi@amd.com )
  Avinash-Chandra Pandey ( avinash-chandra.pandey@amd.com )
  Chandra Kumar Ramasamy ( chandrakumar.ramasamy@amd.com )
  Savan Anadani ( savan.anadani@amd.com )

Signed-off-by: sanadani <savan.anadani@amd.com>

sanadani commented:
/cc @penpornk

sanadani committed:
[AMD-ZENDNN] Minor changes to ensure adherence to coding guidelines
- Variable name changes as per C++ styling.
- Minor comment changes.

Signed-off-by: sanadani <savan.anadani@amd.com>

penpornk commented on tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc:
```diff
@@ -0,0 +1,194 @@
+  VLOG(1) << " TF-EAGER-REWRITE Info: OriginalOp= " << orig_op->Name()
+          << " ZenOp=" << zen_op_name;
+
+  TF_CHECK_OK(SetupNewOp(orig_op, zen_op_name, zen_op));
```
Right now `SetupNewOp` is only used once here. `CreateGenericZenOp` also doesn't do much apart from calling `SetupNewOp`. Do you plan to have another Zen-op creation function that will call `SetupNewOp` in the future? If not, let's just merge `SetupNewOp` and `CreateGenericZenOp` into one function.

sanadani replied:
Accepted the change. Thank you!

penpornk commented on tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc:
```diff
@@ -0,0 +1,194 @@
+}
+
+bool ZenEagerOpRewrite::ShouldRewriteOp(EagerOperation *op) {
+  // Don't rewrite the op if TF-ZenDNN use is disabled at runtime.
+  if (!IsZenDnnEnabled()) {
+    return false;
+  }
```
Let's do this check at the `ZenEagerOpRewrite::Run` level instead.

sanadani replied:
Accepted the change. Thank you!

penpornk reviewed (changes requested):
Thank you very much for the PR! I have minor comments.

sanadani committed:
[AMD-ZENDNN] Refactored CreateGenericZenOp function

- Merged SetupNewOp operations into CreateGenericZenOp function
- Reduced zen_eager_ops_ look up
- Minor comment change

Signed-off-by: sanadani <savan.anadani@amd.com>

sanadani commented on tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc:
```diff
@@ -0,0 +1,194 @@
+  bool is_remote = false;
+  new_zen_op->reset(new tensorflow::EagerOperation(&orig_op->EagerContext()));
+  TF_RETURN_IF_ERROR(new_zen_op->get()->Reset(zen_op_name.c_str(), nullptr,
+                                              is_remote, nullptr));
```
Fixed. Thank you!

penpornk reviewed (approved):
Thank you very much for the quick changes!

sanadani commented on tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc:
```diff
@@ -0,0 +1,194 @@
+      return true;
+    }
+  }
+  return false;
```
Thank you for noticing that and letting us know. Accepted the change.

sanadani commented on tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc:
```diff
@@ -0,0 +1,194 @@
+Status ZenEagerOpRewrite::Run(
+    EagerOperation *orig_op,
+    std::unique_ptr<tensorflow::EagerOperation> *out_op) {
+  if (ShouldRewriteOp(orig_op)) {
```
Accepted the change. Thank you!

sanadani commented on tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc:
```diff
@@ -0,0 +1,194 @@
+
+  // Copy all attributes to the new op.
+  const NodeDef &kOrigNodeDef = orig_op->MutableAttrs()->BuildNodeDef();
+
```
Fixed. Thank you!

penpornk commented:
Got [AMD ROCm test failure](http://ml-ci.amd.com:21096/blue/organizations/jenkins/tensorflow%2Fgithub-prs-upstream-master%2FAMD-ROCm-Community-CI-Build/detail/PR-59314/3/pipeline)
```
ERROR: /workspace/tensorflow/core/common_runtime/eager/BUILD:518:11: no such target '//tensorflow/core/graph:zen_graph_util': target 'zen_graph_util' not declared in package 'tensorflow/core/graph' (did you mean 'mkl_graph_util'?) defined by /workspace/tensorflow/core/graph/BUILD and referenced by '//tensorflow/core/common_runtime/eager:execute'
```
I guess we'll have to wait for #59137 to be merged, sync this PR (if needed), then rerun all the tests.

sanadani commented:
Thank you @penpornk  so much for quick review.

The PR https://github.com/tensorflow/tensorflow/pull/59137 is merged. 
@penpornk @gbaned Can you please re-trigger the CI Builds and other tests for this PR?

Thank you!

penpornk commented:
@sanadani Triggering them now. :) If they still can't find `//tensorflow/core/graph:zen_graph_util`, you might have to rebase the PR. We should know in an hour.

sanadani commented:
Hi @penpornk,

Thank you for triggering run quickly.

It looks like the build was successful. But the wheel file installation [failed ](https://github.com/tensorflow/tensorflow/actions/runs/3994639680/jobs/6852582775#step:5:16835) 

Can you please check? We request for any detail/direction that could help us to know more about this issue and resolve it.

penpornk commented:
@sanadani Thank you for checking! [The Arm CI failure](https://github.com/tensorflow/tensorflow/actions/runs/3994639680/jobs/6852582775) is unrelated to this PR. It failed building `h5py` and not `tensorflow`. 
```
ERROR: Failed building wheel for h5py
```
The Arm CI pipeline has been failing [for the past few days](https://github.com/tensorflow/tensorflow/actions/workflows/arm-ci.yml) so we should just ignore it.

I'm waiting for the 1MacOS CPU Python3` and `Windows Bazel` CI to be done and then I'll import the PR. :)
